### PR TITLE
[GLUTEN-2193][CH] Support slice function

### PR DESCRIPTION
**Pull request by exmy**
## What changes were proposed in this pull request?

`slice(x, start, length)`

There are some differences between Spark and CH on function `slice`:
1. Spark slice only accepts three arguments, but CH can accept two or three;
2. Spark slice returns null when any of the argument is null, but CH applies a specific default behavior when `start` or `length` is null;
3. Spark slice throws exception when `start = 0`  or `length < 0`, but CH not.

## How was this patch tested?

unit tests

### CODE REVIEW

File: cpp-ch/local-engine/CMakeLists.txt [hunk 1]
```diff
@@ -87,11 +87,19 @@ if (ENABLE_LOCAL_UDFS)
     endforeach()
 endif ()
 
-# Notice: soures files under Parser/*_udf subdirectories must be built into target ${LOCALENGINE_SHARED_LIB} directly
+# Add function_parser sources files in sub directories to functions_sources
+file(GLOB children CONFIGURE_DEPENDS RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} Parser/function_parser)
+foreach(child ${children})
+    add_headers_and_sources(function_parsers ${child})
```
Comment (taiyang-li) [inline cpp-ch/local-engine/CMakeLists.txt:93]: Parser/function_parser下还有子目录吗，如果没有的话，直接add_headers_and_sources(function_parsers Parser/function_parser)是不是更简单点

File: cpp-ch/local-engine/Parser/SerializedPlanParser.h [hunk 2]
```diff
@@ -205,6 +205,7 @@ static const std::map<std::string, std::string> SCALAR_FUNCTIONS
        // array functions
        {"array", "array"},
        {"size", "length"},
+       {"slice", ""}, // parser by FunctionParserArraySlice
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Parser/SerializedPlanParser.h:208]: slice已经注册到FunctionParserFactory里了，没必要在这里加。

**Review by taiyang-li — CHANGES_REQUESTED** [review]

File: cpp-ch/local-engine/Parser/function_parser/FunctionParserArraySlice.cpp [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+}
+}
+
+namespace local_engine
+{
+
+class FunctionParserArraySlice : public FunctionParser
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Parser/function_parser/FunctionParserArraySlice.cpp:18]: 文件命名直接用  slice更直观些

**Review by taiyang-li — CHANGES_REQUESTED** [review]

File: cpp-ch/local-engine/CMakeLists.txt (hunk 1 above)
Comment (taiyang-li) [reply cpp-ch/local-engine/CMakeLists.txt:93]: image
将function_parser重命名成scalar_function_parser以复用上面的逻辑

**Review by taiyang-li — CHANGES_REQUESTED** [review]

**Review by taiyang-li — APPROVED** [review]
LGTM

**Review by zzcclp — APPROVED** [review]
LGTM

**Pull request merged**